wallet, rpc: `FundTransaction` refactor

josibake opened the pull request:
## Motivation

The primary motivation for this PR is to enable `FundTransaction` to take a vector of `CRecipient` objects to allow passing BIP352 silent payment addresses to RPCs that use `FundTransaction` (e.g. `send`, `walletcreatefundedpsbt`). To do that, SFFO logic needs to be moved out of `FundTransaction` so the `CRecipient` objects with the correct SFFO information can be created and then passed to `FundTransaction`.

As a secondary motivation, this PR moves the SFFO stuff closer to the caller, making the code cleaner and easier to understand. This is done by having a single function which parses RPC inputs for SFFO and consistently using the `set<int>` method for communicating SFFO.

I'm also not convinced we need to pass a full `CMutableTx` object to `FundTransaction`, but I'm leaving that for a follow-up PR/discussion, as its not a blocker for silent payments.

josibake committed:
refactor: simplify `CreateRecipients`

Move validation logic out of `CreateRecipients` and instead take the
already validated outputs from `ParseOutputs` as an input.

Move SFFO parsing out of `CreateRecipients` into a new function,
`InterpretSubtractFeeFromOutputsInstructions`. This takes the SFFO instructions
from `sendmany` and `sendtoaddress` and turns them into a set of integers.
In a later commit, we will also move the SFFO parsing logic from
`FundTransaction` into this function.

Worth noting: a user can pass duplicate addresses and addresses that dont exist
in the transaction outputs as SFFO args to `sendmany` and `sendtoaddress`
without triggering a warning. This behavior is preserved in to keep this commit
strictly a refactor.

josibake committed:
refactor: pass CRecipient to FundTransaction

Instead turning tx.vout into a vector of `CRecipient`, make `FundTransaction`
take a `CRecipient` vector directly. This allows us to remove SFFO logic from
the wrapper RPC `FundTransaction` since the `CRecipient` objects have already
been created with the correct SFFO values. This also allows us to remove
SFFO from both `FundTransaction` function signatures.

This sets us up in a future PR to be able to use these RPCs with BIP352
static payment codes.

S3RK commented on src/wallet/rpc/spend.cpp:
```diff
@@ -73,6 +73,38 @@ std::set<int> InterpretSubtractFeeFromOutputInstructions(const UniValue& subtrac
     return set_sffo;
 }
 
+std::vector<CRecipient> ParseOutputs(const UniValue& outputs_in, const UniValue& options)
```
Let's not pass all the options as we need only one of them.

josibake replied:
We are using either `subtract_fee_from_outputs` or `subtractFeeFromOutputs`, depending on which one is present. I think handling that logic in the `ParseOutputs` function is better. Since the options object is passed by reference there isn't any performance difference in passing a reference to the entire options object vs passing a single key.

S3RK commented on src/wallet/rpc/spend.cpp:
```diff
@@ -29,16 +29,27 @@ std::vector<CRecipient> ParseRecipients(const UniValue& address_amounts, const s
+        CTxDestination dest;
+        CAmount amount{0};
+        if (address == "data") {
+            if (has_data) {
```
`has_data` never set to `true`

Also no tests fail, so it seems there is no test coverage for that

josibake replied:
Good catch!

S3RK commented on src/wallet/spend.h:
```diff
@@ -224,7 +224,7 @@ util::Result<CreatedTransactionResult> CreateTransaction(CWallet& wallet, const
  * calling CreateTransaction();
  */
-util::Result<CreatedTransactionResult> FundTransaction(CWallet& wallet, const CMutableTransaction& tx, std::optional<unsigned int> change_pos, bool lockUnspents, const std::set<int>& setSubtractFeeFromOutputs, CCoinControl);
+util::Result<CreatedTransactionResult> FundTransaction(CWallet& wallet, const CMutableTransaction& tx, const std::vector<CRecipient>& recipients, std::optional<unsigned int> change_pos, bool lockUnspents, CCoinControl);
```
Not a huge fan of this change. Now if we pass tx.vout they'll be ignored.

I'd prefer if we pass `txin` and `recipients` and don't pass `tx` at all. That would be a clearer interface in my opinion.

josibake replied:
I agree, see my comment in the description. Ultimately, I think `FundTransaction` should take a set of inputs and a set of outputs and return a `CreatedTransactionResult`. #25273 and this PR get us incrementally closer to that, but I don't want to fully tackle that in this PR as it's a bit more complicated than it seems, and the primary goal of this PR is to create the `CRecipients` objects in a standard way across all the RPC calls and pass them to `FundTransaction`

achow101 replied:
I think it would be useful to assert that `tx.vout.empty()` somewhere to make it clear that it is not used. Perhaps this function can do `tx.vout.clear()`, and the `FundTransaction` called by it can do the check.

josibake replied:
Good point; I'll add a check and a comment.

josibake replied:
Rather than clearing tx.vout in the outer RPC function call, I am clearing tx.vout at the call sites of FundTransaction. Otherwise, I would have needed to make `tx` non-const, which seems risky since we are still using it to pass `tx.vins`

S3RK commented on src/wallet/rpc/spend.cpp:
```diff
@@ -532,17 +543,14 @@ static std::vector<RPCArg> FundTxDoc(bool solving_data = true)
 }
 
-CreatedTransactionResult FundTransaction(CWallet& wallet, const CMutableTransaction& tx, const UniValue& options, CCoinControl& coinControl, bool override_min_fee)
+CreatedTransactionResult FundTransaction(CWallet& wallet, const CMutableTransaction& tx, const std::vector<CRecipient>& recipients, const UniValue& options, CCoinControl& coinControl, bool override_min_fee)
```
Same comment as for other `FundTransaction`. I think we should get rid of ambiguity in input parameters and don't pass `tx`. In this case it's even more confusing, because `tx.vout` is actually still checked.

josibake replied:
Also a good catch, we should be using the `recipients` vector to get the size here.

S3RK reviewed (commented):
I'm still trying to fully grasp the existing code (which is a big mess), so bear with me if I don't make sense. Left some comment/questions based on my current understanding.

josibake commented:
Fixed the `has_data` logic and replaced all references to `tx.vout` with `recipients` in `FundTransaction`, per feedback from @S3RK

josibake committed:
refactor: move normalization to new function

Move the univalue formatting logic out of AddOutputs and into its own function,
`NormalizeOutputs`. This allows us to re-use this logic in later commits.

josibake committed:
refactor: move parsing to new function

Move the parsing and validation out of `AddOutputs` into its own function,
`ParseOutputs`. This allows us to re-use this logic in `ParseRecipients` in a
later commit, where the code is currently duplicated.

The new `ParseOutputs` function returns a CTxDestination,CAmount tuples.
This allows the caller to then translate the validated outputs into
either CRecipients or CTxOuts.

josibake committed:
refactor: remove out param from `ParseRecipients`

Have `ParseRecipients` return a vector of `CRecipients` and rename to `CreateRecipients`.

josibake committed:
test: add tests for fundrawtx and sendmany rpcs

If the serialized transaction passed to `fundrawtransaction` contains
duplicates, they will be deserialized and added to the transaction. Add
a test to ensure this behavior is not changed during the refactor.

A user can pass any number of duplicated and unrelated addresses as an
SFFO argument to `sendmany` and the RPC will not throw an error (note,
all the rest of the RPCs which take SFFO as an argument will error if
the user passes duplicates or specifies outputs not present in the
transaction). Add a test to ensure this behavior is not changed during
the refactor.

josibake commented on test/functional/wallet_fundrawtransaction.py:
```diff
@@ -147,6 +150,34 @@ def run_test(self):
+        address = self.nodes[0].getnewaddress("bech32")
+        tx = CTransaction()
+        tx.vin = []
+        tx.vout = [CTxOut(1 * COIN, bytearray(address_to_scriptpubkey(address))) for _ in range(2)]
```
noted! will update

josibake commented on test/functional/wallet_sendmany.py:
```diff
@@ -0,0 +1,43 @@
+
+from test_framework.test_framework import BitcoinTestFramework
+
+class SendmanyTest(BitcoinTestFramework):
```
I initially started putting these tests in `wallet_basic.py`, but I'd prefer if we have RPC-specific test files and move all of the `sendmany` specific tests into the new `wallet_sendmany.py` file in a follow-up PR.

aureleoules commented on src/wallet/rpc/spend.cpp:
```diff
@@ -24,34 +24,16 @@
+    std::vector<CRecipient> recipients;
+    int idx{0};
+    for (const auto& [destination, amount] : outputs) {
+        CRecipient recipient = {destination, amount, subtract_fee_outputs.count(idx) == 1};
```
```suggestion
        CRecipient recipient = {destination, amount, subtract_fee_outputs.contains(idx)};
```

josibake replied:
nice! TIL c++20 introduced contains for sets :heart:

aureleoules commented on src/wallet/rpc/spend.cpp:
```diff
@@ -76,6 +58,50 @@ static void InterpretFeeEstimationInstructions(const UniValue& conf_target, cons
+        subtract_fee_outputs = (options.exists("subtract_fee_from_outputs") ? options["subtract_fee_from_outputs"] : options["subtractFeeFromOutputs"]).get_array();
+        for (const auto& sffo : subtract_fee_outputs.getValues()) {
+            pos = sffo.getInt<int>();
+            if (sffo_set.count(pos))
```
```suggestion
            if (sffo_set.contains(pos))
```

aureleoules commented on src/wallet/rpc/spend.cpp:
```diff
@@ -703,21 +718,10 @@ CreatedTransactionResult FundTransaction(CWallet& wallet, const CMutableTransact
     }
 
-    if (tx.vout.size() == 0)
+    if (recipients.size() == 0)
```
```suggestion
    if (recipients.empty())
```

aureleoules reviewed (commented):
A few nits

josibake commented:
`git range-diff  5892014...909c6bb`

Addresses review comments from @aureleoules and @achow101 , most notable being now clearing `tx.vout` and asserting that it is empty inside both `FundTransaction` functions. This ensures this is no longer used to pass outputs to `FundTransaction`, now and in future code.

josibake commented on src/wallet/rpc/spend.cpp:
```diff
@@ -78,6 +58,36 @@ static void InterpretFeeEstimationInstructions(const UniValue& conf_target, cons
+    int pos{0};
+    std::set<int> sffo_set;
+    UniValue subtract_fee_outputs(UniValue::VARR);
+    if (options.exists("subtractfeefromamount") || options.exists("subtractfeefrom")) {
```
> Since all of the different ways that sffo is set by the user, this could do type interpretation to figure it out.

That makes sense, what do you think about doing the type interpretation on the UniValue passed in, instead of relying on the key names? I'd rather do that than try to have this function take a `std::variant` or something like that

achow101 replied:
I think UniValue type interpretation would be fine.

josibake replied:
Done

josibake commented:
Updated https://github.com/bitcoin/bitcoin/commit/909c6bb1dbc6f60c5ad3200d83f3fce98f4c3706 -> https://github.com/bitcoin/bitcoin/commit/c856a1093009c6098c7c9fed5cc3f03fd0c0cf5b ([compare](https://github.com/josibake/bitcoin/compare/909c6bb1dbc6f60c5ad3200d83f3fce98f4c3706..c856a1093009c6098c7c9fed5cc3f03fd0c0cf5b))

* Moved `CTxDestination`, `Amount` into loop
* Refactored `InterpretSubtractFeeFromAmountInstructions` to infer parsing logic based on the instruction type (int, bool, str)

Thanks for the suggestion, @achow101 , it looks a lot cleaner now.

josibake commented on src/wallet/rpc/spend.cpp:
```diff
@@ -298,12 +303,12 @@ RPCHelpMan sendtoaddress()
-
-    std::vector<CRecipient> recipients = CreateRecipients(address_amounts, subtractFeeFromAmount);
+    UniValue subtractFeeFromAmount(UniValue::VOBJ);
+    subtractFeeFromAmount.pushKV("subtractfeefromamount", request.params[4]);
```
Fixed

josibake commented on src/wallet/rpc/spend.cpp:
```diff
@@ -24,34 +24,14 @@
+    for (const auto& [destination, amount] : outputs) {
+        CRecipient recipient = {destination, amount, subtract_fee_outputs.contains(idx)};
         recipients.push_back(recipient);
+        idx++;
```
Done

josibake commented on src/rpc/rawtransaction_util.cpp:
```diff
@@ -97,41 +97,50 @@ UniValue NormalizeOutputs(const UniValue& outputs_in)
     bool has_data{false};
-
+    CTxDestination destination;
+    CAmount amount;
```
Fixed

S3RK commented on src/wallet/rpc/spend.cpp:
```diff
@@ -78,6 +57,27 @@ static void InterpretFeeEstimationInstructions(const UniValue& conf_target, cons
     }
 }
 
+std::set<int> InterpretSubtractFeeFromOutputInstructions(const UniValue& sffo_instructions, const std::vector<std::string>& destinations)
```
in 0b0276bb00d3511172d662a94ffd7b55adbbf2d1

I don't see a reason to extract this function at this point (could be done later in the PR).
There are two disparate cases covered (`bool` and `vector<string>`). And each case is only used in one and only one place, further commits also don't introduce further usage of this code path as they pass `vector<int>`. 

Wouldn't it be better if we keep this logic at the call site? 
Benefits of my proposal:
* we don't need second parameter for this function
* we don't need to pass dummy value in `fundrawtransaction` RPC code
* we can combine `NormalizeOutputs` and `ParseOutputs`. they are used always together and the only reason for `NormalizeOutputs` to exist now is too pass second param to `InterpretSubtractFeeFromOutputInstructions` in `send` and `walletfundpsbt` RPC. But there second parameter is useless as it can't be in the `vector<string>` type

josibake replied:
I prefer having a single function (as opposed to keeping the logic at the call site) for the following reasons:

* We avoid duplicating code at the RPC call sites:
   * The logic for validating SFFO inputs would need to be duplicated at `send`, `fundrawtx` and `walletcreatefundedpsbt`. This is the code that checks that the ints passed are not duplicated and in bounds and previously used to be inside `rpc/FundTransaction`
* We have one place to validate all SFFO inputs:
  * Currently, `sendmany` SFFO inputs are not validated (documented in the added functional tests)
  * Having one function for parsing and validating makes it cleaner to add validation for all inputs in a follow-up PR. The only reason I'm not doing it in this PR is that its a behavior change (it could make calls to `sendmany` invalid that were previously accepted) and I'd like to keep this strictly a refactor
  
> we don't need second parameter for this function

In the case of `sendmany`, we do need the array of destinations to lookup their position. In the case of `send`, `walletcreatefundedpsbt`, and `fundrawtransaction`, we need the array of destinations to make sure our ints are in bounds. There are definitely other ways to do it, but it seems fine to use the same vector for lookups and size. More just making the point that the argument is always used.

> we can combine NormalizeOutputs and ParseOutputs. they are used always together

Actually, the main goal here was to separate the logic in `AddOutputs` to get rid of duplicate validation logic in `AddOutputs` and also the original `ParseRecipients` function. `sendtoaddress` and `sendtoaddress` need the validation logic, but don't need `NormalizeOutputs` since those RPCs already only allow the user to pass a key-value pair for address and amounts.

The other RPCs (`send`, `fundrawtransaction`, `walletcreatefundedpsbt`) need the logic for parsing how the univalue is passed (`NormalizeOutputs`) and validating the addresses (`ParseOutputs`).

S3RK commented:
Code review ACK c856a1093009c6098c7c9fed5cc3f03fd0c0cf5b

Nice improvements removing duplication between `CreateRecipient` and `AddOutputs`, also embedding SFFO values in `CRecipient` instead of passing as a separate parameter to `FundTransaction`.

I'd still prefer `InterpretSubtractFeeFromOutputInstructions` to work with one type of SFFO option - `vector<int>` and process others at the call site. SFFO option can be in three different forms and `vector<int>` is the only form that used multiple times, the other forms are only used once. But that's not blocking as `InterpretSubtractFeeFromOutputInstructions` and the RPCs are in the same file.

josibake commented:
> According to my `range-diff` nothing changed. reACK [18ad1b9](https://github.com/bitcoin/bitcoin/commit/18ad1b9142e91cef2f5c6a693eeb2d0fbb8c517d)

for context, this was the reason for the rebase: https://github.com/bitcoin/bitcoin/pull/29218#issuecomment-1906227242

achow101 commented on src/rpc/rawtransaction_util.cpp:
```diff
@@ -97,41 +97,49 @@ UniValue NormalizeOutputs(const UniValue& outputs_in)
-
-            CTxOut out(0, CScript() << OP_RETURN << data);
-            rawTx.vout.push_back(out);
+            CTxDestination destination{CNoDestination{CScript() << OP_RETURN << data}};
+            CAmount amount{0};
+            parsed_outputs.emplace_back(destination, amount);
```
In f7384b921c3460c7a3cc7827a68b2c613bd98f8e "refactor: move parsing to new function"

nit: Making temp variables  is not necessary.

achow101 reviewed (commented):
ACK 18ad1b9142e91cef2f5c6a693eeb2d0fbb8c517d
